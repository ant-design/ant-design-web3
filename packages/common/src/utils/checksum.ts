import { isAddress, keccak256, toUtf8Bytes } from 'ethers';

import { ChainType } from '../types';

/**
 * @param address - An address string
 * @desc 将以太坊地址转换为经过校验和处理后的以太坊地址。
 * @returns	The checksum address
 * @example
 * ```ts
 * toChecksumAddressWithEVM('0x237dfeca906991dc570595a8176bfe65be3e8a7e');
 * > "0x237DFECA906991dc570595A8176bfE65Be3E8A7E"
 * ```
 */
export function toChecksumAddressWithEVM(address: string) {
  if (!isAddress(address)) {
    console.error('address is not a valid address', address);
    return address;
  }

  const lowerCaseAddress = address.toLowerCase().replace(/^0x/i, '');
  // 经过 isAddress 校验后 keccak256 不需要判断 EIP-1052
  const hash = keccak256(toUtf8Bytes(lowerCaseAddress));
  let ret = '';

  for (let i = 0; i < address.length; i++) {
    if (parseInt(hash[i], 16) >= 8) {
      ret += address[i].toUpperCase();
    } else {
      ret += address[i];
    }
  }
  return ret;
}

/**
 * @param address - 需要校验和的地址字符串
 * @param chain - 区块链类型
 * @desc 根据不同的区块链类型对地址进行校验和处理。目前支持 EVM 链的校验和地址生成。
 * @returns 经过校验和处理的地址，对于不支持的链类型返回原地址
 * @example
 * ```ts
 * checksumAddress('0x237dfeca906991dc570595a8176bfe65be3e8a7e', ChainType.EVM);
 * > "0x237DFECA906991dc570595A8176bfE65Be3E8A7E"
 *
 * checksumAddress('0x1234567890abcdef', ChainType.SOLANA);
 * > "0x1234567890abcdef" // 返回原地址，因为 Solana 不支持校验和
 * ```
 */
export const checksumAddress = (address: string, chain: ChainType) => {
  switch (chain) {
    case ChainType.EVM: {
      return toChecksumAddressWithEVM(address);
    }
    default:
      console.warn(`checksumAddress: unsupported chain type ${chain}`);
      return address;
  }
};
