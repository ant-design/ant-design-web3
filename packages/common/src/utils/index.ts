export * from './request';
export * from './format';
export * from './test-utils';
export * from './createGetBrowserLink';
export * from './warning';
export * from './checksum';
