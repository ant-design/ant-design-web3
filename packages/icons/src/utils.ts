// for fix https://github.com/ant-design/ant-design-web3/issues/1056
export function eraseDefaultProps(svgComponent: React.FC): React.FC {
  const defaultProps = svgComponent.defaultProps;
  return (props) => {
    const newProps = { ...defaultProps, ...props };
    return svgComponent(newProps);
  };
}
