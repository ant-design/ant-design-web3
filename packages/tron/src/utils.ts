/* v8 ignore start */
import { WalletReadyState } from '@tronweb3/tronwallet-abstract-adapter';

export const hasWalletReady = (readyState?: WalletReadyState) =>
  readyState === WalletReadyState.Found || readyState === WalletReadyState.Loading;
