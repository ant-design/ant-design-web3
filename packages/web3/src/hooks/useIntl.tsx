import React from 'react';
import { ConfigContext } from '@ant-design/web3-common';
import type { ConfigConsumerProps, Locale, RequiredLocale } from '@ant-design/web3-common';

type ComponentName = keyof Locale;

interface IntlType<T extends ComponentName = any> {
  messages: RequiredLocale[T];
  getMessage: (message: string, values?: Record<string, string>) => string;
}

export default function useIntl<T extends ComponentName = any>(
  componentName: T,
  componentLocale?: Locale[T],
): IntlType<T> {
  const context = React.useContext<ConfigConsumerProps>(ConfigContext);
  const locale: RequiredLocale[T] = {
    ...context.defaultLocale[componentName],
    ...context?.locale?.[componentName],
    ...componentLocale,
  };

  return {
    messages: locale,
    getMessage: (message, values) => {
      if (values) {
        return message.replace(/\{([\w\.]+)\}/g, (_, key) => values[key]);
      }
      return message;
    },
  };
}

export type { IntlType };
