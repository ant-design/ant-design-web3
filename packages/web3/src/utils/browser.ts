import copy from 'copy-to-clipboard';

export const getPlatform = () => {
  const userAgent = navigator.userAgent.toLowerCase();
  if (userAgent.includes('chrome')) {
    return 'Chrome';
  }
  if (userAgent.includes('firefox')) {
    return 'Firefox';
  }
  if (userAgent.includes('edge')) {
    return 'Edge';
  }
  if (userAgent.includes('safari')) {
    return 'Safari';
  }
  return 'Other';
};

export const writeCopyText = async (text: string): Promise<boolean> => {
  return copy(text);
};
