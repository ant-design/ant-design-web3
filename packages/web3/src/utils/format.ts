/**
 * @param address address
 * @param prefixProp addressPrefix from props
 * @param prefixContext addressPrefix from context
 */
export const fillWithPrefix = (
  address = '',
  prefixProp?: string | false,
  prefixContext?: string | false,
): string => {
  let prefix = undefined;

  // `0x` is the default prefix
  if (prefixProp === undefined && prefixContext === undefined) {
    prefix = '0x';
  }

  // has prefix be set
  else {
    // use props first
    if (prefixProp !== undefined) {
      prefix = prefixProp === false ? '' : prefixProp;
    }

    // use context
    else if (prefixContext !== undefined) {
      prefix = prefixContext === false ? '' : prefixContext;
    }
  }

  if (!prefix) return address;

  return address.startsWith(prefix) ? address : `${prefix}${address}`;
};

export const formatAddress = (address = '', groupSize = 4): string => {
  const formattedGroups = [];

  const has0x = address.startsWith('0x');
  /* v8 ignore next */
  const formatText = has0x ? address.slice(2) : address;

  for (let i = 0; i < formatText.length; i += groupSize) {
    const group = formatText.slice(i, i + groupSize);
    formattedGroups.push(group);
  }

  const formattedText = formattedGroups.join(' ');

  /* v8 ignore next */
  return has0x ? `0x ${formattedText}` : formattedText;
};

export const formatBalance = (value: bigint | number, decimals: number, fixed?: number): string => {
  const bigValue = typeof value === 'bigint' ? value : BigInt(value);
  const divisor = BigInt(10 ** decimals);
  const displayValue = bigValue / divisor;
  const fraction = bigValue % divisor;

  if (fraction === 0n && fixed === undefined) {
    return `${displayValue}`;
  }

  let fractionStr = fraction.toString().padStart(decimals, '0');
  if (fixed === undefined) {
    return `${displayValue}.${fractionStr.replace(/0+$/, '')}`;
  }
  fractionStr = fractionStr.substring(0, fixed).padEnd(fixed, '0');
  return `${displayValue}.${fractionStr}`;
};
