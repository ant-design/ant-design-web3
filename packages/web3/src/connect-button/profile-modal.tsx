import React, { useContext } from 'react';
import { Address } from '@ant-design/web3';
import type { Balance } from '@ant-design/web3-common';
import { Avatar, Button, ConfigProvider, message, Modal, Space } from 'antd';
import type { AvatarProps, ModalProps } from 'antd';
import classNames from 'classnames';

import { CryptoPrice } from '../crypto-price';
import type { IntlType } from '../hooks/useIntl';
import { writeCopyText } from '../utils';

export interface ProfileModalProps {
  className?: string;
  intl: IntlType;
  /** @internal */
  __hashId__: string;
  avatar?: AvatarProps;
  address?: string;
  name?: string;
  onDisconnect?: () => void;
  open?: boolean;
  onClose?: () => void;
  modalProps?: Omit<ModalProps, 'open' | 'onClose' | 'className'>;
  balance?: Balance;
  addressPrefix?: string | false;
  size?: 'small' | 'middle' | 'large';
}

export const ProfileModal: React.FC<ProfileModalProps> = ({
  className,
  intl,
  __hashId__,
  open,
  onClose,
  onDisconnect,
  avatar,
  name,
  address,
  modalProps,
  balance,
  addressPrefix,
  size,
}) => {
  const { getPrefixCls } = useContext(ConfigProvider.ConfigContext);
  const prefixCls = getPrefixCls('web3-connect-button-profile-modal');
  const [messageApi, contextHolder] = message.useMessage();

  const footer = (
    <div className={classNames(`${prefixCls}-footer`, __hashId__)}>
      {address ? (
        <Button
          size={size}
          onClick={() => {
            writeCopyText(address).then(() => {
              messageApi.success(intl.getMessage(intl.messages.copied));
            });
          }}
        >
          {intl.getMessage(intl.messages.copyAddress)}
        </Button>
      ) : null}
      <Button size={size} onClick={onDisconnect}>
        {intl.getMessage(intl.messages.disconnect)}
      </Button>
    </div>
  );

  return (
    <>
      {contextHolder}
      <Modal
        footer={footer}
        width="fit-content"
        {...modalProps}
        onCancel={onClose}
        className={classNames(className, __hashId__, prefixCls)}
        styles={{
          ...modalProps?.styles,

          body: {
            textAlign: 'center',
            ...modalProps?.styles?.body,
          },
        }}
        open={open}
      >
        <Space align="center" direction="vertical">
          {avatar ? <Avatar {...avatar} /> : null}
          {name ? <div className={classNames(`${prefixCls}-name`, __hashId__)}>{name}</div> : null}
          {balance && <CryptoPrice {...balance} />}
          {address && (
            <Address
              ellipsis={false}
              address={address}
              type={name ? 'secondary' : undefined}
              tooltip={false}
              addressPrefix={addressPrefix}
            />
          )}
        </Space>
      </Modal>
    </>
  );
};
ProfileModal.displayName = 'ProfileModal';
