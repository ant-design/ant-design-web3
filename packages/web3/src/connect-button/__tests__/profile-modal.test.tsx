import { Polygon } from '@ant-design/web3-assets';
import { fireEvent, render } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import { ConnectButton } from '..';
import useIntl from '../../hooks/useIntl';
import { readCopyText } from '../../utils/test-utils';
import { ProfileModal } from '../profile-modal';

describe('ProfileModal', () => {
  it('match snapshot', () => {
    const App = () => {
      const intl = useIntl('ConnectButton');
      return (
        <ProfileModal
          intl={intl}
          open
          __hashId__="hashId"
          address="0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B"
          name="wanderingearth.eth"
          avatar={{
            src: 'https://mdn.alipayobjects.com/huamei_mutawc/afts/img/A*9jfLS41kn00AAAAAAAAAAAAADlrGAQ/original',
          }}
        />
      );
    };
    const { baseElement } = render(<App />);
    expect(baseElement).toMatchSnapshot();
  });

  it('match snapshot without name', () => {
    const App = () => {
      const intl = useIntl('ConnectButton');
      return (
        <ProfileModal
          intl={intl}
          open
          __hashId__="hashId"
          address="0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B"
          avatar={{
            src: 'https://mdn.alipayobjects.com/huamei_mutawc/afts/img/A*9jfLS41kn00AAAAAAAAAAAAADlrGAQ/original',
          }}
        />
      );
    };
    const { baseElement } = render(<App />);
    expect(baseElement).toMatchSnapshot();
  });

  it('open profile modal in ConnectButton', async () => {
    const App = () => (
      <ConnectButton
        account={{
          address: '0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B',
        }}
      />
    );
    const { baseElement } = render(<App />);
    expect(baseElement.querySelector('.ant-web3-connect-button')).not.toBeNull();

    fireEvent.click(baseElement.querySelector('.ant-web3-connect-button')!);

    await vi.waitFor(() => {
      expect(
        baseElement.querySelector('.ant-web3-connect-button-profile-modal .ant-web3-address')
          ?.textContent,
      ).toBe('0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B');
      expect(baseElement.querySelector('.anticon-user')).not.toBeNull();
    });
  });

  it('show balance', async () => {
    const App = () => (
      <ConnectButton
        account={{
          address: '0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B',
        }}
        balance={{
          symbol: 'ETH',
          decimals: 18,
          value: 1230000000000000000n,
        }}
      />
    );
    const { baseElement } = render(<App />);
    expect(baseElement.querySelector('.ant-web3-connect-button')).not.toBeNull();

    fireEvent.click(baseElement.querySelector('.ant-web3-connect-button')!);

    await vi.waitFor(() => {
      expect(
        baseElement.querySelector('.ant-web3-connect-button-profile-modal .ant-web3-address')
          ?.textContent,
      ).toBe('0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B');

      expect(
        baseElement.querySelector(
          '.ant-web3-connect-button-profile-modal .ant-web3-crypto-price-balance',
        )?.textContent,
      ).toBe('1.23 ETH');
    });
  });

  it('show default avatar', async () => {
    const App = () => (
      <ConnectButton
        account={{
          address: '0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B',
        }}
      />
    );
    const { baseElement } = render(<App />);
    expect(baseElement.querySelector('.ant-web3-connect-button')).not.toBeNull();

    fireEvent.click(baseElement.querySelector('.ant-web3-connect-button')!);

    await vi.waitFor(() => {
      expect(
        baseElement.querySelector(
          '.ant-web3-connect-button-chain-icon.ant-web3-connect-button-default-icon',
        ),
      ).not.toBeNull();
    });
  });

  it('show chain icon as the default avatar', async () => {
    const App = () => (
      <ConnectButton
        account={{
          address: '0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B',
        }}
        chain={Polygon}
      />
    );
    const { baseElement } = render(<App />);
    expect(baseElement.querySelector('.ant-web3-connect-button')).not.toBeNull();

    fireEvent.click(baseElement.querySelector('.ant-web3-connect-button')!);

    await vi.waitFor(() => {
      expect(
        baseElement.querySelector(
          '.ant-web3-connect-button-chain-icon .ant-web3-icon-polygon-circle-colorful',
        ),
      ).not.toBeNull();
    });
  });

  it('Disconnect & Copy Address Button', () => {
    const disconnectTestFn = vi.fn();
    const App = () => {
      const intl = useIntl('ConnectButton');
      return (
        <ProfileModal
          open
          intl={intl}
          __hashId__="hashId"
          address="0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B"
          onDisconnect={disconnectTestFn}
        />
      );
    };
    const { baseElement } = render(<App />);
    const btns = baseElement.querySelectorAll('.ant-web3-connect-button-profile-modal .ant-btn');
    expect(btns.length).toBe(2);
    fireEvent.click(btns[1]);
    expect(disconnectTestFn).toBeCalled();
    fireEvent.click(btns[0]);
    expect(readCopyText()).resolves.toBe('0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B');
  });

  it('should not display modal when pass false into profileModal', async () => {
    const App = () => (
      <ConnectButton
        account={{
          address: '0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B',
        }}
        profileModal={false}
      />
    );
    const { baseElement } = render(<App />);
    fireEvent.click(baseElement.querySelector('.ant-web3-connect-button')!);
    await vi.waitFor(() => {
      expect(baseElement.querySelector('.ant-web3-connect-button-profile-modal')).toBeNull();
    });
  });

  it('profile modal should customize by profileModal', async () => {
    const App = () => (
      <ConnectButton
        account={{
          address: '0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B',
        }}
        profileModal={{
          title: 'Custom Title',
          footer: 'Custom Description',
          width: 500,
        }}
      />
    );
    const { baseElement } = render(<App />);
    fireEvent.click(baseElement.querySelector('.ant-web3-connect-button')!);
    await vi.waitFor(() => {
      expect(baseElement.querySelector('.ant-web3-connect-button-profile-modal')).not.toBeNull();
      expect(
        baseElement.querySelector('.ant-web3-connect-button-profile-modal .ant-modal-title')
          ?.textContent,
      ).toBe('Custom Title');
      expect(
        baseElement.querySelector('.ant-web3-connect-button-profile-modal .ant-modal-footer')
          ?.textContent,
      ).toBe('Custom Description');
      expect(
        baseElement
          .querySelector('.ant-web3-connect-button-profile-modal.ant-modal')
          ?.getAttribute('style'),
      ).toContain('width: 500px;');
    });
  });

  it('profile modal should can customize reset styles by `profileModal.styles`', async () => {
    const App = () => (
      <ConnectButton
        account={{
          address: '0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B',
        }}
        profileModal={{
          title: 'Custom Title',
          footer: 'Custom Description',
          width: 500,
          styles: {
            header: {
              fontSize: 32,
            },
            body: {
              textAlign: 'right',
              fontSize: 32,
            },
            footer: {
              fontSize: 32,
            },
          },
        }}
      />
    );

    const { baseElement } = render(<App />);
    fireEvent.click(baseElement.querySelector('.ant-web3-connect-button')!);

    await vi.waitFor(() => {
      const headerStyleAttr = baseElement
        .querySelector('.ant-web3-connect-button-profile-modal .ant-modal-header')
        ?.getAttribute('style');
      const bodyStyleAttr = baseElement
        .querySelector('.ant-web3-connect-button-profile-modal .ant-modal-body')
        ?.getAttribute('style');
      const footerStyleAttr = baseElement
        .querySelector('.ant-web3-connect-button-profile-modal .ant-modal-footer')
        ?.getAttribute('style');

      expect(headerStyleAttr).toContain('font-size: 32px;');

      expect(bodyStyleAttr).toContain('text-align: right;');
      expect(bodyStyleAttr).toContain('font-size: 32px;');

      expect(footerStyleAttr).toContain('font-size: 32px;');
    });
  });
});
