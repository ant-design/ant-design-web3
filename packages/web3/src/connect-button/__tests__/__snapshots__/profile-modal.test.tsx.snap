// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProfileModal > match snapshot 1`] = `
<body>
  <div />
  <div>
    <div
      class="ant-modal-root css-dev-only-do-not-override-ni1kz0"
    >
      <div
        class="ant-modal-mask ant-fade-appear ant-fade-appear-start ant-fade"
      />
      <div
        class="ant-modal-wrap"
        tabindex="-1"
      >
        <div
          aria-modal="true"
          class="ant-modal css-dev-only-do-not-override-ni1kz0 hashId ant-web3-connect-button-profile-modal ant-zoom-appear ant-zoom-appear-prepare ant-zoom"
          role="dialog"
        >
          <div
            style="outline: none;"
            tabindex="0"
          >
            <div
              class="ant-modal-content"
            >
              <button
                aria-label="Close"
                class="ant-modal-close"
                type="button"
              >
                <span
                  class="ant-modal-close-x"
                >
                  <span
                    aria-label="close"
                    class="anticon anticon-close ant-modal-close-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="close"
                      fill="currentColor"
                      fill-rule="evenodd"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M799.86 166.31c.02 0 .04.02.08.06l57.69 57.7c.04.03.05.05.06.08a.12.12 0 010 .06c0 .03-.02.05-.06.09L569.93 512l287.7 287.7c.04.04.05.06.06.09a.12.12 0 010 .07c0 .02-.02.04-.06.08l-57.7 57.69c-.03.04-.05.05-.07.06a.12.12 0 01-.07 0c-.03 0-.05-.02-.09-.06L512 569.93l-287.7 287.7c-.04.04-.06.05-.09.06a.12.12 0 01-.07 0c-.02 0-.04-.02-.08-.06l-57.69-57.7c-.04-.03-.05-.05-.06-.07a.12.12 0 010-.07c0-.03.02-.05.06-.09L454.07 512l-287.7-287.7c-.04-.04-.05-.06-.06-.09a.12.12 0 010-.07c0-.02.02-.04.06-.08l57.7-57.69c.03-.04.05-.05.07-.06a.12.12 0 01.07 0c.03 0 .05.02.09.06L512 454.07l287.7-287.7c.04-.04.06-.05.09-.06a.12.12 0 01.07 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
              <div
                class="ant-modal-body"
                style="text-align: center;"
              >
                <div
                  class="ant-space css-dev-only-do-not-override-ni1kz0 ant-space-vertical ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small"
                >
                  <div
                    class="ant-space-item"
                  >
                    <span
                      class="ant-avatar ant-avatar-circle ant-avatar-image css-dev-only-do-not-override-ni1kz0"
                    >
                      <img
                        src="https://mdn.alipayobjects.com/huamei_mutawc/afts/img/A*9jfLS41kn00AAAAAAAAAAAAADlrGAQ/original"
                      />
                    </span>
                  </div>
                  <div
                    class="ant-space-item"
                  >
                    <div
                      class="ant-web3-connect-button-profile-modal-name hashId"
                    >
                      wanderingearth.eth
                    </div>
                  </div>
                  <div
                    class="ant-space-item"
                  >
                    <span
                      class="ant-typography ant-typography-secondary ant-web3-address css-dev-only-do-not-override-ni1kz0 css-dev-only-do-not-override-ni1kz0"
                    >
                      <span>
                        0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B
                      </span>
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="ant-modal-footer"
              >
                <div
                  class="ant-web3-connect-button-profile-modal-footer hashId"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-ni1kz0 ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
                    type="button"
                  >
                    <span>
                      Copy Address
                    </span>
                  </button>
                  <button
                    class="ant-btn css-dev-only-do-not-override-ni1kz0 ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
                    type="button"
                  >
                    <span>
                      Disconnect
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ProfileModal > match snapshot without name 1`] = `
<body>
  <div />
  <div>
    <div
      class="ant-modal-root css-dev-only-do-not-override-ni1kz0"
    >
      <div
        class="ant-modal-mask ant-fade-appear ant-fade-appear-start ant-fade"
      />
      <div
        class="ant-modal-wrap"
        tabindex="-1"
      >
        <div
          aria-modal="true"
          class="ant-modal css-dev-only-do-not-override-ni1kz0 hashId ant-web3-connect-button-profile-modal ant-zoom-appear ant-zoom-appear-prepare ant-zoom"
          role="dialog"
        >
          <div
            style="outline: none;"
            tabindex="0"
          >
            <div
              class="ant-modal-content"
            >
              <button
                aria-label="Close"
                class="ant-modal-close"
                type="button"
              >
                <span
                  class="ant-modal-close-x"
                >
                  <span
                    aria-label="close"
                    class="anticon anticon-close ant-modal-close-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="close"
                      fill="currentColor"
                      fill-rule="evenodd"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M799.86 166.31c.02 0 .04.02.08.06l57.69 57.7c.04.03.05.05.06.08a.12.12 0 010 .06c0 .03-.02.05-.06.09L569.93 512l287.7 287.7c.04.04.05.06.06.09a.12.12 0 010 .07c0 .02-.02.04-.06.08l-57.7 57.69c-.03.04-.05.05-.07.06a.12.12 0 01-.07 0c-.03 0-.05-.02-.09-.06L512 569.93l-287.7 287.7c-.04.04-.06.05-.09.06a.12.12 0 01-.07 0c-.02 0-.04-.02-.08-.06l-57.69-57.7c-.04-.03-.05-.05-.06-.07a.12.12 0 010-.07c0-.03.02-.05.06-.09L454.07 512l-287.7-287.7c-.04-.04-.05-.06-.06-.09a.12.12 0 010-.07c0-.02.02-.04.06-.08l57.7-57.69c.03-.04.05-.05.07-.06a.12.12 0 01.07 0c.03 0 .05.02.09.06L512 454.07l287.7-287.7c.04-.04.06-.05.09-.06a.12.12 0 01.07 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
              <div
                class="ant-modal-body"
                style="text-align: center;"
              >
                <div
                  class="ant-space css-dev-only-do-not-override-ni1kz0 ant-space-vertical ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small"
                >
                  <div
                    class="ant-space-item"
                  >
                    <span
                      class="ant-avatar ant-avatar-circle ant-avatar-image css-dev-only-do-not-override-ni1kz0"
                    >
                      <img
                        src="https://mdn.alipayobjects.com/huamei_mutawc/afts/img/A*9jfLS41kn00AAAAAAAAAAAAADlrGAQ/original"
                      />
                    </span>
                  </div>
                  <div
                    class="ant-space-item"
                  >
                    <span
                      class="ant-typography ant-web3-address css-dev-only-do-not-override-ni1kz0 css-dev-only-do-not-override-ni1kz0"
                    >
                      <span>
                        0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B
                      </span>
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="ant-modal-footer"
              >
                <div
                  class="ant-web3-connect-button-profile-modal-footer hashId"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-ni1kz0 ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
                    type="button"
                  >
                    <span>
                      Copy Address
                    </span>
                  </button>
                  <button
                    class="ant-btn css-dev-only-do-not-override-ni1kz0 ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
                    type="button"
                  >
                    <span>
                      Disconnect
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
